tsdb/series_segment.go

```diff
@@ -0,0 +1,396 @@
+)
+
+// Series entry constants.
+const (
+	SeriesEntryFlagSize   = 1
+	SeriesEntryHeaderSize = 1 + 8 // flag + id
+
+	SeriesEntryInsertFlag    = 0x01
+	SeriesEntryTombstoneFlag = 0x02
+)
+
+var (
+	ErrInvalidSeriesSegment        = errors.New("invalid series segment")
+	ErrInvalidSeriesSegmentVersion = errors.New("invalid series segment version")
+	ErrSeriesSegmentNotWritable    = errors.New("series segment not writable")
+)
+
+// SeriesSegment represents a log of series entries.
+type SeriesSegment struct {
+	id   uint16
+	path string
+
+	data []byte        // mmap file
+	file *os.File      // write file handle
+	w    *bufio.Writer // bufferred file handle
+	size uint32        // current file size
+}
+
+// NewSeriesSegment returns a new instance of SeriesSegment.
+func NewSeriesSegment(id uint16, path string) *SeriesSegment {
+	return &SeriesSegment{
+		id:   id,
+		path: path,
+	}
+}
+
+// CreateSeriesSegment generates an empty segment at path.
+func CreateSeriesSegment(id uint16, path string) (*SeriesSegment, error) {
+	// Generate segment in temp location.
+	f, err := ioutil.TempFile("", "series-segment-")
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	// Write header to file and close.
+	hdr := NewSeriesSegmentHeader()
+	if _, err := hdr.WriteTo(f); err != nil {
+		return nil, err
+	} else if err := f.Truncate(int64(SeriesSegmentSize(id))); err != nil {
+		return nil, err
+	} else if err := f.Close(); err != nil {
+		return nil, err
+	}
+
+	// Swap with target path.
+	if err := os.Rename(f.Name(), path); err != nil {
+		return nil, err
+	}
+
+	// Open segment at new location.
+	segment := NewSeriesSegment(id, path)
+	if err := segment.Open(); err != nil {
+		return nil, err
+	}
+	return segment, nil
+}
+
+// Open memory maps the data file at the file's path.
+func (s *SeriesSegment) Open() error {
+	if err := func() (err error) {
+		// Memory map file data.
+		if s.data, err = mmap.Map(s.path, int64(SeriesSegmentSize(s.id))); err != nil {
+			return err
+		}
+
+		// Read header.
+		hdr, err := ReadSeriesSegmentHeader(s.data)
+		if err != nil {
+			return err
+		} else if hdr.Version != SeriesSegmentVersion {
+			return ErrInvalidSeriesSegmentVersion
+		}
+
+		return nil
+	}(); err != nil {
+		s.Close()
+		return err
+	}
+
+	return nil
+}
+
+// InitForWrite initializes a write handle for the segment.
+// This is only used for the last segment in the series file.
+func (s *SeriesSegment) InitForWrite() (err error) {
+	// Only calculcate segment data size if writing.
+	for s.size = uint32(SeriesSegmentHeaderSize); s.size < uint32(len(s.data)); {
+		flag, _, _, sz := ReadSeriesEntry(s.data[s.size:])
+		if flag == 0 {
+			break
+		}
+		s.size += uint32(sz)
+	}
+
+	// Open file handler for writing & seek to end of data.
+	if s.file, err = os.OpenFile(s.path, os.O_WRONLY|os.O_CREATE, 0666); err != nil {
+		return err
+	} else if _, err := s.file.Seek(int64(s.size), os.SEEK_SET); err != nil {
```

Comment:
`os.SEEK_SET` is [deprecated](https://golang.org/pkg/os/#pkg-constants), you should use `io.SeekCurrent` instead.

Reply:
Fixed in 3900c94.